velox/dwio/parquet/reader/ParquetReader.cpp

```diff
@@ -506,26 +506,50 @@ ParquetRowReader::ParquetRowReader(
 void ParquetRowReader::filterRowGroups() {
   auto rowGroups = readerBase_->fileMetaData().row_groups;
   rowGroupIds_.reserve(rowGroups.size());
-  auto excluded =
-      columnReader_->filterRowGroups(0, dwio::common::StatsContext());
-
-  for (auto i = 0; i < rowGroups.size(); i++) {
-    VELOX_CHECK_GT(rowGroups_[i].columns.size(), 0);
-    auto fileOffset = rowGroups_[i].__isset.file_offset
-        ? rowGroups_[i].file_offset
-        : rowGroups_[i].columns[0].file_offset;
-    VELOX_CHECK_GT(fileOffset, 0);
-    auto rowGroupInRange =
-        (fileOffset >= options_.getOffset() &&
-         fileOffset < options_.getLimit());
-    // A skipped row group is one that is in range and is in the excluded list.
+
+  auto splitStartOffset = options_.getOffset();
+  auto splitEndOffset = options_.getLimit();
+
+  int64_t lastOffset = 0;
+  for (auto i = 0; i < rowGroups.size() - 1; i++) {
+    auto nextRowGroup = rowGroups_[i + 1];
+    auto flags = nextRowGroup.__isset;
+
+    VELOX_CHECK_GT(nextRowGroup.columns.size(), 0);
+
+    // The file_offset of the RowGroup is optional. When it's not available, we
+    // need to use the ColumnChunk's file_offset instead. A ColumnChunk may not
+    // have statistics but is required to have the file_offset as per the
+    // specification. We use the next RowGroup's first ColumnChunk's file_offset
+    // as the end offset of the current RowGroup.
+    auto startOffset = lastOffset;
+    auto endOffset = flags.file_offset ? nextRowGroup.file_offset
+                                       : nextRowGroup.columns[0].file_offset;
+
+    VELOX_CHECK_GE(startOffset, 0);
+    bool rowGroupInRange =
+        startOffset <= splitEndOffset && endOffset >= splitStartOffset;
+    // A matched row group is one that is overlapping with the split range, and
```

Comment:
Is this accurate? A row group can overlap multiple split ranges (if a split range is small). How will we decide which split range will process a rowGroup?

Reply:
Split is the basic unit of processing. We never process multiple split in the lifetime of the ParquetReader or DwrfReader. See `HiveDataSource::addSplit`, the reader was created new for each split. 
So this problem is to match a split range with multiple RowGroup ranges. For each RowGroup we need to test if it overlaps with the split range.

Reply:
This was actually a mistake. Presto/Velox handles full RowGroups not partial. I have changed it back to just use the RG starting offset.